DOC-662: TinyMCE 5.5 A11ychecker changes (#1699)
tinymce/tinymce-docs: plugins/a11ychecker.md

{{site.premiumplugin}}

The `a11ychecker` premium plugin allows you to check the HTML in the editor for various WCAG & Section 508 accessibility problems. It has an auto-repair feature that lets the user fix identified problems.

## Interactive example

{% include live-demo.html id="a11ychecker" %}

## Basic setup

To add the {{pluginname}} plugin to the editor, add `{{plugincode}}` to the `plugins` option in the editor configuration.

For example:

```js
tinymce.init({
  selector: 'textarea',
  plugins: 'a11ychecker',
  toolbar: 'a11ycheck'
});
```

> ## Accessibility Rules

Author: Added by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

The following checks are performed by the {{pluginname}} plugin.

<a class="anchor" id="D1"></a>

### D1 - Usage of paragraphs as headings

> **Rule description:** This rule checks that `h1`-`h6` tags are used for heading content, not `p` tags. Not using correct heading markup will make it difficult for assistive technologies to visually represent and navigate through your content.

Author: I think it makes sense to just drop 'visually' :)

> **Rule description:** This rule checks that `h1`-`h6` tags are used for heading content, not `p` tags. Not using correct heading markup will make it difficult for assistive technologies to visually represent and navigate through your content.

Author: Fixed by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5
* **WCAG 2.1 specification**: [H48 - Using ol, ul and dl for lists or groups of links](https://www.w3.org/WAI/WCAG21/Techniques/html/H48.html).

<a class="anchor" id="D4U"></a>

### D4U - Unordered list structure

**Rule description:** This rule checks that a `ul` element is used for unordered lists. Do not use paragraphs beginning with `*` or `-` or some similar character instead of an `ol` element containing `li` items. This is to simplify navigation and parsing of the content for users of assistive technologies.

#### {{pluginname}} rule details - D4U

* **Notification level (severity)**: Error
* **[WCAG level](https://www.w3.org/TR/WCAG21/#conformance)**: Level A ; Level AA ; Level AAA
* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5
* **WCAG 2.1 specification**: [H48 - Using ol, ul and dl for lists or groups of links](https://www.w3.org/WAI/WCAG21/Techniques/html/H48.html).

<a class="anchor" id="D5"></a>
<a class="anchor" id="D5A"></a>
<a class="anchor" id="D5B"></a>
<a class="anchor" id="D5C"></a>

### D5 - Contrast ratio of the text (D5A, D5B, and D5C)

> **Rule description:** This rule checks that the contrast ratio of the text is above the specified values (4.5:1 for normal text and 3:1 for large text). Text with a low contrast ratio is hard to read for users with an impaired vision.

Author: Expanded the description in https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

#### {{pluginname}} rule details - D5A

* **Notification level (severity)**: Error
* **[WCAG level](https://www.w3.org/TR/WCAG20/#conformance)**: Level AA
* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5
* **WCAG 2.0 specification**: [Ensuring that a contrast ratio of at least 3:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G145.html).

#### {{pluginname}} rule details - D5B

* **Notification level (severity)**: Error
* **[WCAG level](https://www.w3.org/TR/WCAG20/#conformance)**: Level AA
* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5
* **WCAG 2.0 specification**: [Ensuring that a contrast ratio of at least 4.5:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G18.html).

#### {{pluginname}} rule details - D5C

* **Notification level (severity)**: Error
* **[WCAG level](https://www.w3.org/TR/WCAG20/#conformance)**: Level AAA
* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5

> * **WCAG 2.0 specification**: [Ensuring that a contrast ratio of at least 7:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G17.html).

Author: Re 'prefix': Looking at this, I think I forgot I was doing it part way down the page... 
Re 'link': Better to be specific. Those general pages can be discussing many issues, most of which we don't or can't test.

<a class="anchor" id="H93"></a>

### H93 - IDs must be unique

**Rule description:** This rule checks that all `id` attributes are unique in the editor. Duplicate `id` attributes are known to cause problems for assistive technologies when parsing the content.

#### {{pluginname}} rule details - H93

* **Notification level (severity)**: Error
* **[WCAG level](https://www.w3.org/TR/WCAG21/#conformance)**: Level A ; Level AA ; Level AAA
* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5
* **WCAG 2.1 specification**: [H93 - Ensuring that id attributes are unique on a Web page](https://www.w3.org/WAI/WCAG21/Techniques/html/H93.html).

<a class="anchor" id="I1"></a>

> ### I1 - Image ALT text

Author: Fixed by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

**Rule description:** This rule checks that all images have alternative (`alt`) text for screen readers and other assistive technologies.

#### {{pluginname}} rule details - I1

* **Notification level (severity)**: Error
* **[WCAG level](https://www.w3.org/TR/WCAG21/#conformance)**: Level A ; Level AA ; Level AAA
* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5
* **WCAG 2.1 specification**: [G95 - Providing short text alternatives that provide a brief description of the non-text content](https://www.w3.org/WAI/WCAG21/Techniques/general/G95.html).

<a class="anchor" id="I2"></a>

> ### I2 - Alt text filename

Author: Replaced by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

#### {{pluginname}} rule details - I2

* **Notification level (severity)**: Error
* **[WCAG level](https://www.w3.org/TR/WCAG21/#conformance)**: Level A ; Level AA ; Level AAA
* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5
* **WCAG 2.1 specification**: [G95 - Providing short text alternatives that provide a brief description of the non-text content](https://www.w3.org/WAI/WCAG21/Techniques/general/G95.html).

<a class="anchor" id="T1"></a>

### T1 - Table caption

**Rule description:** This rule checks that all `table` elements have a `caption` element describing the data inside the table to simplify parsing and navigation of the content for users of assistive technologies.

#### {{pluginname}} rule details - T1

* **Notification level (severity)**: Error
* **[WCAG level](https://www.w3.org/TR/WCAG21/#conformance)**: Level A ; Level AA ; Level AAA
* **[HTML version](#a11ychecker_html_version)**: HTML4 and HTML5
* **WCAG 2.1 specification**: [H39 - Using caption elements to associate data table captions with data tables](https://www.w3.org/WAI/WCAG21/Techniques/html/H39.html).

<a class="anchor" id="T2"></a>

### T2 - Complex table summary

> **Rule description:** This rule checks that all complex tables must have a `summary` attribute explaining to users of assistive technologies how to navigate through the data inside of the table.

Author: Added by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

#### {{pluginname}} rule details - T2

* **Notification level (severity)**: Warning
* **[WCAG level](https://www.w3.org/TR/WCAG21/#conformance)**: Level A ; Level AA ; Level AAA
* **[HTML version](#a11ychecker_html_version)**: HTML4
* **WCAG 2.1 specification**: [H73 - Using the summary attribute of the table element to give an overview of data tables](https://www.w3.org/WAI/WCAG21/Techniques/html/H73.html).

<a class="anchor" id="T3"></a>

### T3 - Table caption and summary

> **Rule description:** This rule checks that the table caption and summary does not have the same text content. The caption should explain **what** the table is about while the summary should explain **how** to navigate the data inside of the table.

Author: Added by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

**Possible Values:** `true`, `false`

#### Example: Using `a11ychecker_allow_decorative_images`

```js
tinymce.init({
  selector: 'textarea',
  plugins: 'a11ychecker',
  toolbar: 'a11ycheck',
  a11ychecker_allow_decorative_images: true
});
```

### `a11ychecker_filter_issue`

> The `a11ychecker_filter_issue` option allows Accessibility Checker rules to be removed from the Accessibility Checker report using a callback function. This option is a flexible alternative to the `a11ychecker_ignored_rules` option.

Reviewer: ```suggestion
The `a11ychecker_filter_issue` option allows Accessibility Checker issues to be removed from the Accessibility Checker report using a callback function. This option is a flexible alternative to the `a11ychecker_ignored_rules` option.
```
since it specifically let's you remove violations in each report based on whatever info you want (e.g. rule id, element, etc.), not just removing specific rules?

> The function will be passed the result object of the [`getReport()` API](#getreport). To filter issues from the Accessibility Checker report, the callback needs to return `false` for the issue.

Reviewer: "result object"?

Reviewer: ```suggestion
The function will be passed the result object of the [`getReport()` API](#getreport). To filter an issue from the Accessibility Checker report, the callback needs to return `false` for that particular issue.
```
maybe, for clarity?

Reviewer: this also runs when you open the dialog too (since it calls the `getReport` API)

**Type:** `Function`

**Default Value:**

```js
function(issue) {
  return true;
}
```

> #### Example: Using `a11ychecker_filter_issue` filter the Accessibility Checker T1 rule

Reviewer 2: ```suggestion
#### Example: Using `a11ychecker_filter_issue` to filter out the Accessibility Checker T1 rule
```

The callback function in the following example will return `false` if the issue `id` value is `'T1'`, filtering `'T1'` issues from the Accessibility Checker report.

```js
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: 'a11ychecker',
  toolbar: 'a11ycheck',
  a11ychecker_filter_issue: function(issue) {
    return issue.id !== 'T1';
  }
});
```

> #### Example: Using `a11ychecker_filter_issue` filter all Accessibility Checker table rules

Reviewer 2: ```suggestion
#### Example: Using `a11ychecker_filter_issue` to filter (out?) all Accessibility Checker table rules
```

> The callback function in the following example will return `false` if the issue `element` value is `'table'`, filtering all table-related issues from the Accessibility Checker report.

Reviewer: ```suggestion
The callback function in the following example will return `false` if the issue `element` value is `'table'`, filtering all `table` element-related issues from the Accessibility Checker report.
```
Since this would specifically do only table elements, and not issues that affects tds, etc.?

```js
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: 'a11ychecker',
  toolbar: 'a11ycheck',
  a11ychecker_filter_issue: function(issue) {
    return issue.element.nodeName.toLowerCase() !== 'table';
  }
});
```

### `a11ychecker_html_version`

This configuration option sets the HTML version to use when checking issues.

> For example: Setting the version to HTML 4 will trigger the rule "Complex tables should have summaries", where `summary` is a valid attribute in HTML 4 but not HTML 5.

Reviewer 2: I kind of prefer the level of detail in the old explanation. However, either way we should probably change it from "will trigger the rule" to some variation of either "will allow the rule to be triggered" or "will trigger the rule on [example HTML]"

Author: "will allow the rule to be triggered" is a passive-voice version of the current change, which we do not want ;)
to improve clarity, I'll replace "trigger" with "enable"

Author: ```suggestion
For example: Setting the version to HTML 4 will enable the rule "Complex tables should have summaries", where `summary` is a valid attribute in HTML 4 but not HTML 5.
```

> For example: Setting the version to HTML 4 will enable the rule "Complex tables should have summaries", where `summary` is a valid attribute in HTML 4 but not HTML 5.

Reviewer: where/because? not sure which is better here

Author: I have an irrational dislike of "because", so where?

**Type:** `String`

**Default value:** `html4`

**Possible Values:** `html4`, `html5`

#### Example: Using `a11ychecker_html_version`

```js
tinymce.init({
  selector: 'textarea',
  plugins: 'a11ychecker',
  toolbar: 'a11ycheck',
  a11ychecker_html_version: 'html5'
});
```

### `a11ychecker_ignored_rules`

> The `a11ychecker_ignored_rules` option prevents specific Accessibility Checker rules being checked. This feature allows developers to skip rules that flag known issues in content until the issue is fixed.

Reviewer: ```suggestion
The `a11ychecker_ignored_rules` option prevents specific Accessibility Checker rules from being checked. This feature allows developers to skip rules that they do not wish to check. For example: to skip rules that flag known content issues or custom HTML that should not be checked.
```
maybe??

> **Type:** `String`

Author: ```suggestion
**Type:** A comma-separated string.
```

**Default Value:** `''`

> **Possible values:** `'D1'`, `'D2'`, `'D3'`, `'D4O'`, `'D4U'`, `'D5A'`, `'D5B'`, `'D5C'`, `'H93'`, `'I1'`, `'I2'`, `'T1'`, `'T2'`, `'T3'`, `'T4A'`, `'T4B'`, `'T4C'`

Reviewer 2: Should we instead say that the possible values are a comma / space separated list containing {...}?

Author: yes but not here. I'll add this information to "Type"

Author: See: https://github.com/tinymce/tinymce-docs/pull/1699/files#diff-16872d5ac386c71e74962d4d3ff8dc6eR382

#### Example: Using `a11ychecker_ignored_rules`

This examples shows how to ignore the following checks (rules):

* D2 - Sequential headings

> * I2 - Alt text filename

Reviewer: If we change I2 name above, change here too

Author: Fixed by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

* T4B - Table headers

```js
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: 'a11ychecker',
  toolbar: 'a11ycheck',
  a11ychecker_ignored_rules: 'D2, I2, T4B'
});
```

### `a11ychecker_issue_url_callback`

The `a11ychecker_issue_url_callback` option is used to change the target URL for the "Click for more info" button (![help icon - a question mark inside a circle]({{site.baseurl}}/images/icons/help.svg)) in the Accessibility Checker dialog. By default, the "more info" links will point to `{{site.url}}/docs/plugins/a11ychecker/#<ruleId>`, such as `{{site.url}}/docs/plugins/a11ychecker/#D1`. This option can be used to set the target URL to a page or pages outside {{site.url}}/docs/.

**Type:** `Function`

> **Default Value:** `undefined`

Reviewer 2: Should we tell a small lie here and say that the default value is a function that returns `{{site.url}}/docs/plugins/a11ychecker/#${ruleId}` (as even if it might not be done through a default parameter, that's the default behaviour when you don't set it)? Or do you think the example use below provides enough of an example for the style of function to write here

Author: > By default, the "more info" links will point to `{{site.url}}/docs/plugins/a11ychecker/#<ruleId>`, such as `{{site.url}}/docs/plugins/a11ychecker/#D1`

Already stated in the description to reduce confusion a bit (I hope)

#### Example: Using `a11ychecker_issue_url_callback`

This example shows how to change the link for the "Click for more info" button (![help icon - a question mark inside a circle]({{site.baseurl}}/images/icons/help.svg)) on the Accessibility Checker dialog to point to anchors at `www.example.com/tinymce/a11ychecker/more_info`.

```js
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  plugins: 'a11ychecker',
  toolbar: 'a11ycheck',
  a11ychecker_issue_url_callback: function (ruleId) {
    return `https://www.example.com/tinymce/a11ychecker/more_info#${ruleId}`
  };
});
```

### `a11ychecker_level`

This configuration option sets the [WCAG level](https://www.w3.org/TR/WCAG20/#conformance) to use when checking issues.

For example, the "Text must have a contrast ratio of at least ..." rule:

> * When using **AA**, Accessibility Checker will check for a contrast ratio of less than 4.5:1.

Reviewer 2: To me, "check for a contrast ratio of $X" would read as the plugin trying to verify that the contrast ratio _is_ $X, and erroring if it is _not_ $X. I might be wrong though, feel free to ignore

Author: fair, I'll reword

Author: Fixed by a366c514e3559853e05910bd21715ecf47024b8a

* When using **AAA**, Accessibility Checker will check that the contrast ratio is not less than 7.0:1.

**Type:** `String`

**Default value:** `aa`

**Possible Values:** `a`, `aa`, `aaa`

#### Example: Using `a11ychecker_level`

```js
tinymce.init({
  selector: 'textarea',
  plugins: 'a11ychecker',
  toolbar: 'a11ycheck',
  a11ychecker_level: 'aaa'
});
```

{% include misc/plugin-toolbar-button-id-boilerplate.md %}

{% include misc/plugin-menu-item-id-boilerplate.md %}

## APIs

The {{pluginname}} plugin provides the following APIs.

### `toggleaudit()`

> Opens the accessibility checker dialog with the results of the audit and options to correct the problems, if any.

Reviewer: Do we care to mention that if the dialog is open, calling this closes it?

Author: Fixed by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798

#### Example: Using `toggleaudit()`

```js
editor.plugins.a11ychecker.toggleaudit();
```

### `getReport()`

> Conducts accessibility audit and reports the results without triggering the dialog. The report represents an array of issues, with the following details:

Reviewer 2: ```suggestion
Conducts accessibility audit and reports the results without triggering the dialog. The report represents an array of issues, each with the following details:
```

Author: ```suggestion
Conducts accessibility audit and reports the results without triggering the dialog. The report produced is an array of objects, with each object representing an issue and contains the following details:
```
@ashfordneil ?

Reviewer 2: containing?

Author: no, then it sounds like the issue contains those details, not the object.

Author: Fixed by https://github.com/tinymce/tinymce-docs/pull/1699/commits/36bc08703715c08ebab95f6f83d6b60cc1daac3e

> Conducts accessibility audit and reports the results without triggering the dialog. The report produced is an array of objects, where each object represents an issue and contains the following details:

Reviewer: ```suggestion
Conducts an accessibility audit and reports the results without opening the dialog. The report produced is an array of objects, where each object represents an issue and contains the following details:
```

* **id** - The issue identifier used by {{site.productname}}, such as D1, T4A. For a description and other details about the issue, see [Accessibility Rules](#accessiblityrules).
* **description** - A description of the issue.
* **severity** - The severity of the issue, either: `info`, `warning`, or `error`.
* **element** - The DOM element where the issue was found.
* **url** - A URL to the W3 WCAG technique that needs to be addressed to clear the issue.

#### Example: Using `getReport()`

```js
var issues = editor.plugins.a11ychecker.getReport();

console.log(issues);

// example result
[
  {
    "id": "D2",
    "severity": "error",
    "url": "http://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html",
    "description": "Headings must be applied in sequential order. For example: Heading 1 should be followed by Heading 2, not Heading 3.",

>     "element": {}  // The element value contains the DOM element (<hn>).

Reviewer 2: ```suggestion
    "element": {}  // The element value contains the DOM element (eg. <h4>).
```
Could we do something like this instead? I was close to googling this mysterious `hn` element I hadn't seen before to find out what it was.

Author: kinda, can't accept that suggestion, we don't use i.e. or e.g. I'll make a similar change

Author: ```suggestion
    "element": {}  // The element value contains the DOM element (such as <h4>).
```

  },
  {
    "id": "T1",
    "severity": "error",
    "url": "http://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html",
    "description": "Tables must have captions",

>     "element": {}  // The element value contains the DOM element (<table>).

Reviewer: do we care that we have `such as` on the `element` description above but not here?

Author: Fixed by https://github.com/tinymce/tinymce-docs/pull/1699/commits/57201f8a1b8a53d3c0861a819b68e45f66e68798
